Add Natural Resources Wales integration to monitor river levels in Wales, UK

gimpje opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change adds sensors for the various river level monitoring stations around Wales. The data is provided by [Natural Resources Wales](https://naturalresources.wales/evidence-and-data/maps/check-river-levels/?lang=en).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
sensor:
  - platform: natural_resources_wales
    api_key: YOUR_API_KEY_HERE
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13223

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

gimpje committed:
initial attempt at sensor

gimpje committed:
refactor to support the 3 sensor types. just stubs for flood risk and flood warnings so far

gimpje committed:
remove unimplemented sensors

gimpje committed:
increase default polling interval to 60 secs

gimpje committed:
update manifest

frenck commented on homeassistant/components/natural_resources_wales/river_levels_sensor.py:
```diff
@@ -0,0 +1,153 @@
+"""Classes for Natural Resources Wales river levels."""
```
This should be extracted to a separate package published on PyPi, see:

https://developers.home-assistant.io/docs/creating_component_code_review/#4-communication-with-devicesservices

frenck commented on homeassistant/components/natural_resources_wales/sensor.py:
```diff
@@ -0,0 +1,50 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
We don't allow for configuration via a platform integration anymore and require configuring an integration via a configuration flow.


For more information, check out ADR-0007 & ADR-0010:

- https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md
- https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md

For the developer documentation on building a configuration flow see:

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

frenck reviewed (commented):
Hi there @gimpje!

Thanks for opening a PR, left two somewhat larger comments that need to be addressed before we can continue on the review process.

Could you take a look? Thanks! 👍

gimpje commented:
Hi @frenck - thanks for the review.

I've just moved house (got flooded last February, hence the river level integration :p) but I'll try to get these changes made soon.

vcc-sanya committed:
Update lib zigpy-cc to fix issues (#35364)

frenck committed:
Merge branch 'dev' into rc

frenck committed:
Bumped version to 0.110.0b0

bramkragten committed:
Updated frontend to 20200514.0 (#35598)

hunterjm committed:
Additional checks for ONVIF event capabilities (#35599)

catch any exceptions when pulling event capabilities and assume it is not supported

bouwew committed:
Bump haanna to 0.15.0 (#35579)

* Link to haanna v0.15.0

* Update requirements_all.txt

zacpotts committed:
Fix zwave thermostat specific device type (#35609)

marcelveldt committed:
Fix zwave_mqtt creating the device name (#35603)

* Fix for creating the device name

Creating the devicename included a typo and was missing the custom (preferred) name set by the OZW Admin tool.

* update comments

scop committed:
Upgrade huawei-lte-api to 1.4.12 (#35618)

https://github.com/Salamek/huawei-lte-api/releases/tag/1.4.12

amelchio committed:
Upgrade to pysonos 0.0.29 (#35617)

hunterjm committed:
additional log info and strings fix (#35622)

bdraco committed:
Fix reversed logic in zeroconf homekit pairing check (#35596)

* Fix reversed logic in zeroconf homekit pairing check

* s/server_info/service_info/

bramkragten committed:
Add check for HTML in translations (#35615)

* Add check for HTML in translations

and remove existing html

* Add test

uvjustin committed:
Clean up forked_daapd volume saving/setting in async_play_media (#35584)

* Clean up volume saving/setting in async_play_media

* Set source to pipe when queued externally

* Add server version requirement to error string

Adminiuga committed:
Don't remove deprecated ZHA config option yet (#35627)

bramkragten committed:
Updated frontend to 20200514.1 (#35632)

frenck committed:
Rename zwave_mqtt to ozw (#35631)

StevenLooman committed:
Properly handle incomplete upnp ssdp discovery (#35553)

frenck committed:
Update translations for OZW and ONVIF

frenck committed:
Bumped version to 0.110.0b1

Quentame committed:
Bump python-synology to 0.8.1 (#35640)

* Bump python-synology to 0.8.1

* Fix tests

bramkragten committed:
Bumped version to 0.110.0b2

shenxn committed:
Bump roombapy to 1.6.1 (#35650)

* Bump roombapy to 1.6.1

* Improve roomba error handling

bramkragten committed:
Fix caldav event for calendar panel (#35653)

bdraco committed:
Update cast to use shared zeroconf (#35570)

* Update cast to use the shared zeroconf instance

* Add zeroconf to after_dependencies

* Bump version to 5.2.0

hunterjm committed:
ONVIF: Add check around media capabilities (#35667)

bramkragten committed:
Updated frontend to 20200515.0 (#35677)

uvjustin committed:
Change MediaPlayerDevice to MediaPlayerEntity (#35692)

Commit:
Update Universal Powerline Bus event name (#35644)

ctalkington committed:
Prevent discovery of IPP printers lacking identifier (#35630)

frenck committed:
Bumped version to 0.110.0b3

frenck committed:
Fix handling of additional data in core config storage (#35660)

Adminiuga committed:
Bump up ZHA dependencies (#35706)

Danielhiversen committed:
Upgrade opengarage lib to 0.1.4 (#35729)

uvjustin committed:
Abort forked-daapd zeroconf flow if version < 27 (#35709)

* Change MediaPlayerDevice to MediaPlayerEntity

* Abort zeroconf if mtd-version < 27.0

bdraco committed:
Ensure homekit version strings conform to spec (#35741)

HomeKit requires all version strings to be in the
format MAJOR.MINOR.REVISION

Danielhiversen committed:
Update mill manifest to reflect config flow (#35748)

bdraco committed:
Handle UPS disconnects in NUT (#35758)

fredrike committed:
Fix daikin discovery flow (#35767)

MatsNl committed:
Bump Atag dependency to 0.3.1.2 (#35776)

bramkragten committed:
Updated frontend to 20200518.0 (#35785)

uvjustin committed:
Skip forked_daapd ignored entries with empty entry.data (#35772)

frenck committed:
Bumped version to 0.110.0b4

hunterjm committed:
Fix ONVIF subscription renewal (#35792)

* fix subscription renewal

* catch ValueError for #35762

amelchio committed:
Upgrade pysonos to 0.0.30 (#35793)

Adminiuga committed:
Bump up ZHA dependencies. (#35797)

uvjustin committed:
Change version check in forked-daapd zeroconf step (#35796)

bramkragten committed:
Updated frontend to 20200519.0 (#35813)

bramkragten committed:
Bumped version to 0.110.0b5

Kane610 committed:
UniFi - Fix disabled entities being enabled after a restart (#35819)

* Async remove call removed too much, resulting in disabled entities coming back after a restart

* Calling super().async_remove is no longer needed, changed to self.async_remove

* Yes, they should be sets...

frenck committed:
Bumped version to 0.110.0

frenck committed:
Merge pull request #35828 from home-assistant/rc

fredrike committed:
Fix Daikin duplicate entries (#35833)

Emilv2 committed:
Fix Delijn sensor naming (#35789)

uvjustin committed:
Handle None received from pyforked-daapd (#35830)

* Handle None received from API in forked-daapd

* Bump pyforked-daapd version in requirements

* Add test

bachya committed:
Fix Prezzibenzina doing I/O in the event loop (#35881)

* Fix Prezzibenzina doing I/O in the event loop

* Linting

hunterjm committed:
fix mjpeg issue along with some cameras not returning event capabilities properly (#35885)

bdraco committed:
Ensure http can startup if homekit fails to load (#35888)

* Ensure HomeAssistant can startup if homekit fails to load

* Update homeassistant/components/logbook/manifest.json

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Fix legacy Hunter Douglas PowerView devices (#35895)

These devices are missing firmware information
as the 1.0 firmware did not provide it.

ctalkington committed:
Update rokuecp to 0.4.1 (#35899)

* update rokuecp to 0.4.1

* Update requirements_all.txt

* Update requirements_test_all.txt

bramkragten committed:
Updated frontend to 20200519.1 (#35877)

bdraco committed:
Ensure storage write consume the data under the lock (#35889)

If two writes trigger at the same time the data would already
be consumed.

bdraco committed:
Homekit should skip devices that are missing in device registry (#35857)

* Homekit should skip devices that are missing in device registry

* Add test for this failure state

danielperna84 committed:
Fix light profiles for HomeMatic lights (#35882)

pvizeli committed:
Bump version 0.110.1

bachya committed:
Auto-level AirVisual API calls (#34903)

pvizeli committed:
Merge pull request #35907 from home-assistant/rc

0.110.1

bramkragten committed:
Update frontend to 20200519.3 (#35925)

* Updated frontend to 20200519.2

* Updated frontend to 20200519.3

fredrike committed:
Fix Daikin AC integration for AirBase units (#35952)

Commit:
Fix typo in conf[CONF_SSL] (#35946)

hunterjm committed:
Bump pyAV and close unclosed outputs (#35960)

* bump pyAV and close unclosed outputs

* skip stream from coverage for now

* fix divide by zero error

bramkragten committed:
Update frontend to 20200519.4 (#35987)

MartinHjelmare committed:
Bump python-openzwave-mqtt to 1.0.2 (#35980)

MartinHjelmare committed:
Fix lutron_caseta setup options (#35974)

hunterjm committed:
Fix ONVIF Transport (#35932)

* allow lib to create AsyncTransport

* fix transport close issue

bdraco committed:
Ensure homekit functions if numpy is unavailable (#35931)

hunterjm committed:
Fix ONVIF PTZ and profile encoding issues (#36006)

* allow lib to create AsyncTransport

* fix transport close issue

* fix zoom only cameras without PTZ presets

* catch profiles without encoding configuration

* also catch ServerDisconnectedError for ptz

flz committed:
Fix iaqualink sensors (#36000)

* iaqualink: small sensor fixes

* Re-add device_class, fix type hints.

flz committed:
Bump iaqualink to 0.3.3 (#35999)

alengwenus committed:
Fix device_registry cleanup behavior (#35977)

* Fix: Only decives which are not referenced by an entity or a config_entry are removed

* Adapted test for async_cleanup

* Changed variable names

ctalkington committed:
Fix roku play/pause support (#35991)

felipediel committed:
Handle StorageError in the Broadlink integration (#35986)

soldag committed:
Migrate rpi_gpio_pwm to extend LightEntity instead of Light (#36028)

bdraco committed:
Use a single service browser for zeroconf discovery (#35997)

chmielowiec committed:
Fix service registration supported features check (#35718)

bdraco committed:
Fix shade compatibility with hunter douglas powerview 1.0 hubs (#36040)

frenck committed:
Upgrade hass-nabucasa to 0.34.3 (#36025)

Jc2k committed:
Bump aiohomekit (#36041)

ctalkington committed:
Improve ipp unique id parsing (#35959)

fredrike committed:
Bump tellduslive version (#36048)

frenck committed:
Bumped version to 0.110.2

Danielhiversen committed:
Fix opengarage async_setup_platform (#36075)

* opengarage async_setup_platform

* async_add_entities

frenck committed:
Merge pull request #36070 from home-assistant/rc

hunterjm committed:
Fix ONVIF config entry unique ID (#36008)

* fallback to device serial number if no mac available

* make password optional to fix #35904

* update tests to reflect new flow

* fix snake case and AsyncMock

* add comments around why weird things are being done

Julius2342 committed:
Bump pyvlx to 0.2.16 (#35971)

hunterjm committed:
guard against missing topic (#36108)

hunterjm committed:
fix preset warning (#36110)

ctalkington committed:
Update rokuecp to 0.4.2 (#36102)

* update rokuecp to 0.4.2

* Update requirements_all.txt

* Update requirements_test_all.txt

ctalkington committed:
Fix roku play/pause during standby (#36096)

Minims committed:
Fix onvif snapshot for Sricam SP009 (#36095)

isk0001y committed:
Re-read last imap_email_content email when no change (#36065)

bdraco committed:
Remove unsupported stop feature with Hunter Douglas Powerview 1.0 Hubs (#36129)

bdraco committed:
Ensure homekit bridge state is restored before creating devices (#36098)

* Ensure homekit bridge state is restored before creating devices

* Tests to ensure homekit device registry entry is stable

* remove stray continue

balloob committed:
Fix client ID lookup for official apps (#36131)

Tho85 committed:
Fix emulated_hue compatibility with older devices (#36090)

* Fix emulated_hue compatibility with older devices

* Fix test ugliness

* Fix pylint errors

bachya committed:
Fix bugs with AirVisual auto-leveling API (#36097)

* Fix bugs with AirVisual auto-leveling API

* Code review

* Code review

bramkragten committed:
Update frontend to 20200519.5 (#36154)

frenck committed:
Bumped version to 0.110.3

frenck committed:
Merge pull request #36155 from home-assistant/rc

Commit:
Fix cloud connection within API (#36158)

gadgetchnnel committed:
Check todoist due date is not None in async_get_events (#36140)

* Check that due date is not None

Check that due date is not None, prevents taks without due dates from breaking Calendar API

* Invert None check to reduce indentation

emontnemery committed:
Let PAHO MQTT client handle connection to MQTT server (#35983)

* Let PAHO client handle connection to MQTT server

bdraco committed:
Add ability to ignore heos discovery (#34653)

* Add ability to ignore heos discovery

* Fetch player_id, update tests

* Handle failure state

* Update tests as there are two players in the mock now

* Adjust and add more tests

* Strip out player id lookup

* reverts per review

* one more revert

JeffLIrion committed:
Fix Android TV icon when screencap option is disabled (#35710)

* Don't return a media image hash if the screencap config option is False

* 1-liner

hunterjm committed:
Fix empty preset element in ONVIF response (#36182)

balloob committed:
Revert DSMR not calling entity methods (#36179)

emontnemery committed:
Bump zeroconf, pychromecast. Log if zeroconf.get_service_info fails. (#36185)

ctalkington committed:
Fix roku select source with app ids (#36191)

bachya committed:
Prevent AirVisual from polling (#36199)

* Prevent AirVisual from polling

* Docstring

emontnemery committed:
Fix custom position range (#36222)

emontnemery committed:
Correct MQTT device trigger reconfiguration with same topic (#36234)

balloob committed:
Bumped version to 0.110.4

balloob committed:
Merge pull request #36238 from home-assistant/rc

Commit:
Merge branch 'master' into natural_resources_wales

Jc2k commented:
@gimpje I recently added something simlilar to HA that uses the Environment Agency API after a similiar experience myself. But to my disappointment I've recently found out that it doesn't really help Welsh users at all (naive me thinking a .uk API would cover the frigging UK). So if you decided to look at this again I know from tickets im triaging that there are other users that would appreciate it! And I can try and help if you want help with config flow or anything.

gimpje commented:
@Jc2k hi :wave:  - I'm still not home from the flooding in February :cry: :cloud_with_rain: plus with Covid and a busy period at work I've not really had a chance (or the energy :p) to revisit this yet to look at the changes frenck requested.

It doesn't look like I'll be home until mid-November at the earliest so if you wanted to look at this then feel free! But yeah, if not, I'll give you a shout once I'm back on my feet to see what the best way to finish this off will be. :+1:

Jc2k commented:
@gimpje thanks for the update! i'm not in a position to fully take on any more integrations right now (super busy too!) i hope things get back to normal for you soon! When you do come back to this though my Environment Agency integration is [here](https://github.com/home-assistant/core/tree/dev/homeassistant/components/eafm) - if you haven't worked with the config flow API you might be able to copy and adapt the config flow module from it.
